### DISCUSSION TITLE
JSON as a message is not escaped/formatted correctly?

### DISCUSSION THREAD

**Discussion by persona94** [gabime/spdlog#2965] [General]
I set up my formatter with a custom pattern

```
formatter->add_flag<ThreadIdFlags>('j').set_pattern("{\"@timestamp\":\"%Y-%m-%dT%H:%M:%S.%fZ\",\"message\":\"%v\",\"level\":\"%l\",%j,\"source\":\"%s\",\"function\":\"%!\",\"line\":\"%#\"}");
spdlog::set_formatter(std::move(formatter));

```
My flag "j" adds a certain uid that has relevance to my code

While calling the log, I use nlohmann::json to generate a json

```
nlohmann::json j;
  
j["k1"] = "v1";
  
j["k2"] = 2;
 
LOG_ERROR(j.dump()); //LOG_ERROR is SPDLOG_ERROR

``` 

The output json is invalid in the "message" field:

`{"@timestamp":"2023-12-27T15:46:41.636829Z","message":"{"k1":"v1","k2":2}","level":"error","uid":25,"source":"main.cpp","function":"testPost","line":"23"}`

It would work if the string supplied to message were escaped but nlohmann::json doesn't have that capability because `.dump()` prints the object per json spec. Any suggestions on what can be done here? Should I write a custom print for my json obj that returns the json as a string that has the key-value pairs without dquotes?

**Comment by tt4g** [comment]
Do you want to output escaped JSON for the `message`?
If you want the entire log message to be in JSON format, wouldn't it be odd to output escaped JSON only for the `message`?

If you wanted to output a JSON object in the `message` property, just replace `"%v"` with `%v`.

See also: #1797, #2592, #2861

**Comment by persona94** [comment]
The message could be anything - string, integer, pointers, or a json blob/object. We would like to see all messages enclosed by dquotes so I have to go with `"%v"`. If I do that, and don't escape a json object (i.e. the `message` field  then the entire json (i.e. the full log) will be an invalid json. 

Note: Neither LOG_INFO(json.dump()) nor LOG_INFO("{}", json.dump()) escape the message

**Reply by tt4g** [reply]
spdlog does not support the JSON format and does not have a JSON escaping API.
If you want to print escaped JSON, it is your responsibility to implement the escaping process.

**Reply by persona94** [reply]
Fair enough. Thanks
